### DISCUSSION TITLE
Many "Content Security Policy (CSP)" issues while browsing

### DISCUSSION THREAD

**Discussion by DavidGarciaCat** [firefly-iii/firefly-iii#7554] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [x] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [ ] ... there are no open or closed issues that are related to my problem
- [x] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

## Brave Browser

Browsing **any** page
```
fused to load the script 'https://.../cdn-cgi/scripts/5c5dd728/cloudflare-static/email-decode.min.js' because it violates the following Content Security Policy directive: "script-src 'unsafe-eval' 'strict-dynamic' 'self' 'unsafe-inline' 'nonce-iPHdaFiUroWbiWMTHCtFaw=='". Note that 'strict-dynamic' is present, so host-based allowlisting is disabled. Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
```

Trying to enable 2FA:
```
fused to send form data to 'https://.../profile/enable2FA' because it violates the following Content Security Policy directive: "form-action 'self'".
```

During the registration process, I got am email (so the SMTP settings are operational) but I didn't get a redirection. I didn't think the CSP headers could be a problem.

## Mozilla Firefox

Right after login:

```
Content Security Policy: The page's settings blocked the loading of a resource at https://.../cdn-cgi/scripts/5c5dd728/cloudflare-static/email-decode.min.js ("script-src"). [...](https://.../)
    on_content_end TMExt.startRate is undefined [on_content_end.js:34:11](moz-extension://cc27d07c-1211-47e6-93ee-023952c9786b/on_content_end.js)
Content Security Policy: The page's settings blocked the loading of a resource at http://.../json/intro/index ("connect-src"). [app.js:2:119918](https://.../v1/js/app.js?v=6.0.10)
Content Security Policy: The page's settings blocked the loading of a resource at http://.../chart/account/frontpage ("connect-src"). [app.js:2:119918](https://.../v1/js/app.js?v=6.0.10)
Content Security Policy: The page's settings blocked the loading of a resource at http://.../chart/account/expense ("connect-src"). [app.js:2:119918](https://.../v1/js/app.js?v=6.0.10)
Content Security Policy: The page's settings blocked the loading of a resource at http://.../chart/account/revenue ("connect-src"). [app.js:2:119918](https://.../v1/js/app.js?v=6.0.10)
Content Security Policy: The page's settings blocked the loading of a resource at http://.../json/frontpage/piggy-banks ("connect-src"). [app.js:2:119918](https://.../v1/js/app.js?v=6.0.10)
```

### Debug information

Debug information generated at 2023-05-26 00:00:00 Europe/Madrid for Firefly III version **6.0.10**.

| Scope | Version |
| --- | --- |
| Firefly III | 6.0.10 |
| Firefly III API | 2.0.1 |
| Build | 479, 13-05-2023 06:42:29 CEST |
| Base Build | 3, 29-04-2023 14:45:54 CEST |
| DB version | 19 (exp. 19) |
| PHP | `8.2.5` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | Europe/Madrid |
| Browser TZ | Europe/Madrid |
| App environment | production |
| App debug mode | false |
| App cache driver | file |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Max upload | 67108864 (64 MB) |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Trusted proxies (.env) | `` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start |  |
| Session end |  |
| Session first |  |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36 |

### Expected behaviour

A review of the CSP headers to make sure all of them are working and not causing a blocker

### Steps to reproduce

1. Configure the `docker-compose.yaml` file (including the importer) as per the official docs
2. Enable a public DNS (at least, this is what triggered my issue)
3. Try to create an account, and browse clicking on several links and tabs

### Additional info

Searching for existing CSP issues before opening this one. It seems to be two issues:
- https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue+csp+is%3Aclosed
- https://github.com/firefly-iii/firefly-iii/issues/5493
- https://github.com/firefly-iii/firefly-iii/issues/3459

Although they both are closed now, the CSP issue is still happening.
I don't know if the problem is the latest version of this PHP tool is not yet deployed as a docker container

I didn't see any related information in the documentation page, only links to these issues I mentioned above

**Comment by JC5** [comment]
Check out https://github.com/orgs/firefly-iii/discussions/7293

**Reply by DavidGarciaCat** [reply]
Hello @JC5 and thank you for your reply.

I have read the discussion at https://github.com/orgs/firefly-iii/discussions/7293 as per your suggestion.

I don't remember enabling any rules by default, but I have applied the custom rule as per the discussion you suggested.

<img width="727" alt="Screenshot 2023-05-27 at 16 20 25" src="https://github.com/firefly-iii/firefly-iii/assets/7809429/c4d530fe-f31f-4774-9146-edeb627cd8a5">

Still, the CSP headers are throwing errors in my console: so far, i have tried to login and to enable the 2FA, both causing errors in my browser console.

```
Refused to send form data to 'https://.../login' because it violates the following Content Security Policy directive: "form-action 'self'".
```

Please note after getting this error I have browsed the root page to get access to the dashboard, then browse the `Options / Profile / 2FA` menu to get access to that feature and get the following error

```
Refused to send form data to 'https://.../profile/enable2FA' because it violates the following Content Security Policy directive: "form-action 'self'".
```

I didn't try any other feature, because if I can't deal with the authentication, authorisation and 2FA logic, then I don't think it's a good idea to keep adding additional data.

I thought the CSP is a set of rules added by the Laravel framework here. Do you really think the issue is on the Cloudflare side?

**Reply by JC5** [reply]
The demo site is also running on Cloudflare, it works fine. I'm not really sure what the cause could be. What is the exact CSP header you get back?

**Reply by DavidGarciaCat** [reply]
@JC5 according to my browser's console when browsing the dashboard:

```

Refused to connect to '<URL>' because it violates the following Content Security Policy directive: "connect-src 'self'".

app.js?v=6.0.10:2 Refused to connect to 'http://.../json/intro/index' because it violates the following Content Security Policy directive: "connect-src 'self'".

send @ app.js?v=6.0.10:2
app.js?v=6.0.10:2 Refused to connect to 'http://.../chart/account/frontpage' because it violates the following Content Security Policy directive: "connect-src 'self'".

send @ app.js?v=6.0.10:2
app.js?v=6.0.10:2 Refused to connect to 'http://.../chart/account/expense' because it violates the following Content Security Policy directive: "connect-src 'self'".

send @ app.js?v=6.0.10:2
app.js?v=6.0.10:2 Refused to connect to 'http://.../chart/account/revenue' because it violates the following Content Security Policy directive: "connect-src 'self'".

send @ app.js?v=6.0.10:2
app.js?v=6.0.10:2 Refused to connect to 'http://.../json/frontpage/piggy-banks' because it violates the following Content Security Policy directive: "connect-src 'self'".
```

These are the errors when I try to set the 2FA:

```
profile:1 Refused to send form data to 'https://.../profile/enable2FA' because it violates the following Content Security Policy directive: "form-action 'self'".
```

I don't know if there's anything I need to set in my Nginx proxy or if there are additional settings to be included in any other software or service.

What I have is:

```
CloudFlare => Nginx Reverse Proxy => Firefly III Docker Container
```

**Reply by JC5** [reply]
https://docs.firefly-iii.org/data-importer/faq/installation-errors/#i-get-page-load-errors-because-the-protocols-dont-match

**Comment by idoodler** [comment]
I just set it up, disabled the two CloudFlare features and I am still unable to login using a Chrome based browser.

The following error is presented in the Browser console: `Refused to send form data to 'https://.../login' because it violates the following Content Security Policy directive: "form-action 'self'".`

`TRUSTED_PROXIES` is set to `**`
`CSP Header`: `default-src 'none'; object-src 'none'; script-src 'unsafe-eval' 'strict-dynamic' 'self' 'unsafe-inline' 'nonce-NONCE'   ; style-src 'unsafe-inline' 'self'; base-uri 'self'; font-src 'self' data:; connect-src 'self' ; img-src data: 'strict-dynamic' 'self' *.tile.openstreetmap.org ; manifest-src 'self'; form-action 'self'`

**Reply by JC5** [reply]
https://docs.firefly-iii.org/data-importer/faq/installation-errors/#i-get-page-load-errors-because-the-protocols-dont-match

**Comment by dyskette** [comment]
I just want to add that for podman I had to recreate (stop, remove and run again) the container after changing `TRUSTED_PROXIES` in my .env file. It worked fine after that without changes in nginx proxy manager.

**Comment by Dinth** [comment]
Hmm, I'm experiencing a similar problem. Funnily, everything was fine until I moved my Firefly instance to a different internal IP (of course updated the nginx config to correctly point to the correct IP), but after the move all the graphs stopped showing up.

The problem persists after disabling those two CloudFlare features, or even disabling proxied access (set to "only DNS") whatsover
